File: code/ex276.cpp
class Solution {
public:
    int numWays(int n, int k) {
        if (n == 0)    return 0;
        if (n == 1)    return k;
        if (n == 2)    return k * k;
        int prev = k * k, prev2 = k;
        for (int i = 2; i < n; i++) {
            int temp = prev;
            prev = (k - 1) * (prev + prev2);
            prev2 = temp;
        }
        return prev;
    }
};

File: code/ex276.java
class Solution {
    public int numWays(int n, int k) {
        if (n == 0)    return 0;
        if (n == 1)    return k;
        
        int dp1 = k * k, dp11 = k;
        for (int i = 2; i < n; i++) {
            int cur = (k - 1) * (dp1 + dp11);
            dp11 = dp1;
            dp1 = cur;
        }
        
        return dp1;
    }
}